archive/100-theorems-list/83_friendship_graphs.lean

```diff
@@ -0,0 +1,285 @@
+-/
+import combinatorics.adj_matrix
+import linear_algebra.char_poly.coeff
+import data.int.modeq
+import data.zmod.basic
+import tactic
+
+/-!
+# The Friendship Theorem
+
+## Definitions and Statement
+- A `friendship` graph is one in which any two distinct vertices have exactly one neighbor in common
+- A `politician`, at least in the context of this problem, is a vertex in a graph which is adjacent
+  to every other vertex.
+- The friendship theorem (Erdős, Rényi, Sós 1966) states that every finite friendship graph has a
+  politician.
+
+## Proof outline
+The proof revolves around the theory of adjacency matrices.
+- Assume `G` is a finite friendship graph.
+- First we show that any two nonadjacent vertices have the same degree
+- Assume for contradiction that `G` does not have a politician.
+- Conclude from the last two points that `G` is `d`-regular for some `d : ℕ`.
+- Show that `G` has `d ^ 2 - d + 1` vertices.
+- By casework, show that if `d = 0, 1, 2`, then `G` has a politician.
+- If `3 ≤ d`, let `p` be a prime factor of `d - 1`.
+- If `A` is the adjacency matrix of `G` with entries in `ℤ/pℤ`, we show that `A ^ p` has trace `1`.
+- This gives a contradiction, as `A` has trace `0`, and thus `A ^ p` has trace `0`.
+
+-/
+
+open_locale classical big_operators
+noncomputable theory
+
+open finset simple_graph matrix
+
+universes u v
+variables {V : Type u} {R : Type v} [semiring R]
+
+section friendship_def
+variables (G : simple_graph V)
+
+/--
+This is the set of all vertices mutually adjacent to a pair of vertices.
+-/
+def common_friends [fintype V] (v w : V) : finset V := G.neighbor_finset v ∩ G.neighbor_finset w
+
+/--
+This property of a graph is the hypothesis of the friendship theorem:
+every pair of vertices has exactly one common friend.
+-/
+def friendship [fintype V] : Prop := ∀ ⦃v w : V⦄, v ≠ w → (common_friends G v w).card = 1
+
+/--
+A politician is a vertex that is adjacent to all other vertices.
+-/
+def exists_politician : Prop := ∃ (v : V), ∀ (w : V), v ≠ w → G.adj v w
+
+lemma mem_common_friends [fintype V] {v w : V} {a : V} :
+  a ∈ common_friends G v w ↔ G.adj v a ∧ G.adj w a :=
+by simp [common_friends]
+
+end friendship_def
+
+variables [fintype V] [inhabited V] {G : simple_graph V} {d : ℕ} (hG : friendship G)
+include hG
+
+variables (R)
+theorem friendship_adj_sq_apply_of_ne {v w : V} (hvw : v ≠ w) :
+  ((G.adj_matrix R) ^ 2) v w = 1 :=
+begin
+  rw [pow_two, ← nat.cast_one, ← hG hvw],
+  simp [common_friends, neighbor_finset_eq_filter, finset.filter_filter, finset.filter_inter],
+end
+
+lemma friendship_adj_cube_apply_of_not_adj {v w : V} (non_adj : ¬ G.adj v w) :
+  ((G.adj_matrix R) ^ 3) v w = degree G v :=
+begin
+  rw [pow_succ, mul_eq_mul, adj_matrix_mul_apply],
+  dunfold degree, rw card_eq_sum_ones, rw sum_nat_cast,
+  apply sum_congr rfl,
+  intros x hx, rw [friendship_adj_sq_apply_of_ne, nat.cast_one], apply hG,
+  intro contra, rw contra at hx, apply non_adj, rw mem_neighbor_finset at hx, apply hx,
```

Comment:
We generally prefer one tactic call per line unless they're very short/related. This applies in various places in this file.